### PR TITLE
General optimization

### PR DESCRIPTION
**Pull request by workofart**
## Description

1. Move the general data loading to a streaming/mmap-friendly flow.
   - `BytePairEncoder` can train from an iterable of documents instead of requiring one giant string
   - Encoded causal-LM token data is written as `.npy` and can be loaded with `mmap_mode="r"`
   - GPT-2 examples can reuse cached `.npy` token data without fetching or re-tokenizing raw text
   - This opens up the possibility to train with OpenWebText 38GB+ on my local Mac by streaming into memory instead of loading it all at once into memory

2. Move dataset/collator boundaries toward numpy data until batch emission.
   - Datasets and collators do slicing/padding/windowing in numpy
   - The key boundary is `collator` is responsible for converting numpy mmapped arrays into backend-device-specific `xp.array()` batches
   - Update SFT and GRPO batch paths to follow the same numpy data boundary.
     - SFT variable-length prompt/loss-mask examples go through dynamic batch padding.
     - GRPO rollout batches keep generated-token masks and advantages as batch data owned by the collator/
trainer boundary.

3. Add document-boundary handling for causal-LM tokenization.
   - Examples now wrap training documents with `<SOS>` and `<|endoftext|>` consistently.
   - This avoids merging unrelated documents together in the token stream.
   - The tokenizer will need to be re-trained with the new vocabulary for lossless tokenization
   - This is also a prerequisite for proper SFT post-training. Previously, we're just using arbitrary "User: " and "Assistant: " string templates

## Tests

Updated unit tests.

## Performance Comparison

**Tokenizer benchmark**

Synthetic local corpus, same input on main and this branch.

| Case | `main` | This branch |
|---|---|---|
| Tokenizer training wall time | 0.0309s | 0.0116s |
| Tokenizer training docs/sec | 19.4k | 51.7k |
| Encode/write wall time | 0.1008s | 0.0609s |
| Encode/write tokens/sec | 89.3k | 147.8k |
| Peak RSS in combined benchmark | 85.9 MB | 63.5 MB |

**DataLoader benchmark**

Synthetic token stream, `TokenWindowMapDataset` + `CausalLMWindowCollator`.

| Case | Result |
|---|---|
| in-memory (`main` branch) sequential | 31.7M tokens/sec |
| in-memory (`main` branch) random replacement | 29.5M tokens/sec |
| memmap sequential | 32.3M tokens/sec |
| memmap random replacement | 29.5M tokens/sec |

Correctness checks:
- batch input/label shift correctness: true
- loss_total_weight correctness: true

### CODE REVIEW

File: autograd/data/collator.py
Comment (workofart) [inline autograd/data/collator.py:1]: The major change in this `collator.py` file is we're following the contract of:
- input: np.ndarray (mmapped, streamed from disk)
- output: xp.array (backedn-device-specific array, which could be on Numpy, Cupy or MLX)

You might see some functions returning `np.ndarray` as output, that's because those are intermediate functions, the collator contract is enforced at the `Collator.__call__` boundary.

File: autograd/data/dataset.py [hunk 1]
```diff
@@ -59,36 +63,48 @@ class PairedMapDataset(MapDataset):
...
         ...     target_key="loss_mask",
-        ...     dtype=xp.int32,
         ... )
         >>> example = next(iter(causal_lm))
         >>> example["tokens"].tolist(), example["loss_mask"].tolist()
         ([10, 11, 12], [0, 1, 1])
     """
 
+    @staticmethod
+    def _load(data: Union[Sequence[Any], str]) -> Any:
```
Comment (workofart) [inline autograd/data/dataset.py:73]: Datasets are loaded as memory-mapped(mmap), so they're streamed from disk instead of loading all data directly to memory. We still support the previous data structures of loading all into memory, that's the `isinstance(data, np.ndarray)` branch.

File: autograd/data/sft.py
Comment (workofart) [inline autograd/data/sft.py:1]: These are just following the same data loading contract of np.arrays before collator, and xp.arrays after collator.__call__ into batches.

File: autograd/text/tokenizer.py [hunk 2]
```diff
@@ -24,17 +25,30 @@ class BytePairEncoder:
     Examples:
         >>> raw_text = "Hello world! This is a test."
         >>> bpe = BytePairEncoder(num_merges=50)
-        >>> encoded_array = bpe.prepare_data(raw_text) # Outputs an array of token IDs
+        >>> encoded_array = bpe.prepare_data([raw_text]) # Outputs an array of token IDs
     """
 
-    SPECIAL_TOKENS = ["<|endoftext|>", "<PAD>", "<SOS>", "<UNK>"]
+    SPECIAL_TOKENS = [
```
Comment (workofart) [inline autograd/text/tokenizer.py:31]: Adding more special tokens so that we prepare for proper SFT training later. Will need to re-pre-train after vocabulary is re-built

File: autograd/text/tokenizer.py [hunk 3]
```diff
@@ -156,97 +186,180 @@ def _construct_unicode_to_int_vocab(self) -> Dict[bytes, int]:
...
             >>> print(encoded_array)
             [ ...some token IDs... ]
         """
-        # 1) Train the vocabulary if needed
-        self.train_vocabulary(raw_text, overwrite_saved_file=overwrite_vocabulary_file)
+        self.train_vocabulary(texts, overwrite_saved_file=overwrite_vocabulary_file)
+        self.encode_to_mmap(texts, overwrite_encoded_data=overwrite_encoded_data)
+        return np.load(self.mmap_path, mmap_mode="r")
+
+    def encode_to_mmap(
```
Comment (workofart) [inline autograd/text/tokenizer.py:218]: major memory savings by encoding by memory-mapping (streaming from disk)

File: autograd/tools/trainer.py [hunk 4]
```diff
@@ -528,7 +528,7 @@ def _loss_total_weight(self, _batch_data) -> Array:
         return xp.array(1.0, dtype=xp.float32)
 
     @abstractmethod
-    def _compute_loss(self, batch_data) -> Tensor:
+    def _forward_and_loss(self, batch_data) -> Tensor:
```
Comment (workofart) [inline autograd/tools/trainer.py:531]: Renaming to `forward_and_loss` is clearer, while the responsibility previously was already running forward and computing the loss. The motivation for this function is to provide a way for deeper optimizations like fusing the forward call and the loss function call together in a fused kernel.

File: examples/gpt_2.py [hunk 5]
```diff
@@ -246,10 +249,11 @@ def sample(self, model: GPT2, input_ids: Array) -> Tensor:
         # Double-check whether we want to overwrite the encoded_data and vocabulary
         custom_bpe=CustomBpeConfig(
             num_merges=3000,
-            encoded_data_path="training_data/bpe_3000_shakespeare_encoded_data.npz",
+            encoded_data_path="training_data/bpe_3000_shakespeare_bos_eos_encoded_data.npz",
```
Comment (workofart) [inline examples/gpt_2.py:252]: had to re-encode the training data after we updated the tokenizer special vocabuarly

File: examples/gpt_2.py [hunk 6]
```diff
@@ -180,15 +183,15 @@ def forward(self, x: Tensor) -> Tensor:
         a = self.layer_norm1(x)
 
         x = x + self.multi_head_attention(a, a, a, is_causal=True)
-        if low_precision_input:
+        if low_precision_input and x.data.dtype != input_dtype:
```
Comment (workofart) [inline examples/gpt_2.py:186]: Fix a slient data type promotion bug.

E.g.

```python
# x starts as bfloat16
x = Tensor(..., dtype=bfloat16)

# attention internally uses a float32 mask
attn_out = attention(x)        # returns float32 by accident/path promotion

# residual add now promotes x too
x = x + attn_out               # x becomes float32
```

Now the next layer receives float32 instead of bfloat16, so low-precision training silently stops being low
precision for that residual stream.

So x is cast back to bfloat16 only when it actually drifted.
